subsys/net/lib/sockets/sockets.c

```diff
@@ -1221,6 +1221,16 @@ static inline ssize_t zsock_recv_stream(struct net_context *ctx,
 				timeout = Z_TIMEOUT_TICKS(remaining);
 			}
 		}
+
+		if (!K_TIMEOUT_EQ(timeout, K_NO_WAIT) && waitall &&
```

Comment:
The idea seems fine, but I wonder if the logic could be refined a bit to make it less complex. Currently we call `wait_data()` before entering the loop and at the end of the loop. I think it'd be more reasonable to call the `wait_data()` at the beginning of the loop instead, i. e.  add
```
if (!K_TIMEOUT_EQ(timeout, K_NO_WAIT) {
	res = wait_data(ctx, &timeout);
	if (res < 0) { 
		errno = -res; 
		return -1; 
	}
}
```
after `if (sock_is_eof(ctx))` check at line 1147.

Reply:
I tried that, but it did not work (unit tests started to fail). I can investigate more what was the reason.

Reply:
> I tried that, but it did not work (unit tests started to fail). I can investigate more what was the reason.

My earlier experiment with this was clearly wrong as now, when trying it again, it works just fine. Will send a new version shortly.